xyzcrush/src/xyzcrush.cpp

```diff
@@ -0,0 +1,161 @@
+/*
+ * This file is part of xyzcrush. Copyright (c) 2017 xyzcrush authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <zlib.h>
+#include <cstring>
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <vector>
+#ifdef _WIN32
+# include <algorithm>
+#endif
+#include "zopfli/zlib_container.h"
+
+# ifdef __MINGW64_VERSION_MAJOR
+int _dowildcard = -1; /* enable wildcard expansion for mingw-w64 */
+# endif
+
+/** Returns the filename (without extension). */
+std::string GetFilename(const std::string& str);
+
+/** Returns that path (everything left to the last /) */
+std::string GetPath(const std::string& str);
+
+std::string GetFilename(const std::string& str) {
+	std::string s = str;
+#ifdef _WIN32
+	std::replace(s.begin(), s.end(), '\\', '/');
+#endif
+
+	// Extension
+	size_t found = s.find_last_of(".");
+	if(found != std::string::npos) {
+		s = s.substr(0, found);
+	}
+
+	// Filename
+	found = s.find_last_of("/");
+	if(found == std::string::npos) {
+		return s;
+	}
+
+	s = s.substr(found + 1);
+	return s;
+}
+
+std::string GetPath(const std::string& str) {
+	std::string s = str;
+#ifdef _WIN32
+	std::replace(s.begin(), s.end(), '\\', '/');
+#endif
+	// Path
+	size_t found = s.find_last_of("/");
+	if(found == std::string::npos) {
+		return ".";
+	}
+
+	s = s.substr(0, found);
+	return s;
+}
+
+int main(int argc, char* argv[]) {
+	ZopfliOptions zopfli_options;
+	ZopfliInitOptions(&zopfli_options);
+	zopfli_options.verbose = 0;
+	zopfli_options.verbose_more = 0;
+	zopfli_options.numiterations = 15;
+	zopfli_options.blocksplitting = 1;
+	zopfli_options.blocksplittinglast = 0;
+	zopfli_options.blocksplittingmax = 15;
+
+	if (argc < 2) {
+		std::cout << "Usage: " << argv[0]
+			<< " filename" << std::endl;
+		return 1;
+	}
+
+	for (int arg = 1; arg < argc; arg++) {
+		std::ifstream file(argv[arg],
+			std::ios::binary | std::ios::ate);
+		if (!file) {
+			std::cerr << "Error reading file " << argv[arg] << "."
+				<< std::endl;
+			return 1;
+		}
+
+		long size = file.tellg();
+		char* header = new char[4];
+
+		file.seekg(0, std::ios::beg);
+		file.read((char*) header, 4);
+		if (memcmp(header, "XYZ1", 4) != 0) {
+			std::cerr << "Input file " << argv[arg]
+				<< " is not a XYZ file: '" << header << "'."
```

Comment:
This prints more than 4 bytes because it prints until it encounters a null-byte. Maybe make header 5 bytes and set last one to zero.